### PR TITLE
Decompile field_tasks

### PR DESCRIPTION
**Pull request by mrgriffin**
Replaces `asm/field_tasks.s` and `data/field_tasks.s` with `src/field_tasks.c`.
Renames various functions to match pokeemerald.

`sIcefallCaveIceTileCoords` should probably be an array of structs rather than a 2D array of `u8`s, but the obvious approach of `struct { u8 x; u8 y; }` didn't work.

### CODE REVIEW

File: data/specials.inc [hunk 1]
```diff
@@ -317,7 +317,7 @@ gSpecials:: @ 815FD60
 	def_special Special_DrawElevatorCurrentFloorWindow
 	def_special nullsub_75
 	def_special Special_CheckPartyPokerus
-	def_special sub_806E99C
+	def_special SetIcefallCaveCrackedIceMetatiles
```
Comment (jiangzhengwenjz) [inline data/specials.inc]: add `Special_` prefix for functions whose pointer is in special table

File: src/field_tasks.c [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+                data[9] = y;
+            }
+        }
+    }
+}
+
+static void sub_806ED38(void)
+{
+    FlagSet(0x829);
+    FlagSet(0x828);
```
Comment (jiangzhengwenjz) [inline src/field_tasks.c]: check include/constants/flags.h for corresponding macro to use

File: src/field_tasks.c [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+            FlagSet(i + 1);
+            break;
+        }
+    }
+}
+
+void SetIcefallCaveCrackedIceMetatiles(void)
+{
+    u8 i = 0;
+    for (; i < ARRAY_COUNT(sIcefallCaveIceTileCoords); ++i)
```
Comment (jiangzhengwenjz) [inline src/field_tasks.c]: for every `ARRAY_COUNT` in this file, use `NELEMS` instead

File: src/field_tasks.c [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+    sPerStepCallbacks[idx](taskId);
+}
+
+static void Task_RunTimeBasedEvents(u8 taskId)
+{
+    s16 *data = gTasks[taskId].data;
+
+    if (!ScriptContext2_IsEnabled())
+    {
+        if ((u8)(gUnknown_203ADFA - 2) > 1)
```
Comment (jiangzhengwenjz) [inline src/field_tasks.c]: `if (gUnknown_203ADFA != 2 && gUnknown_203ADFA != 3)`

File: src/field_tasks.c [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+        case 2:
+            if (data[6] != 0)
+            {
+                data[6]--;
+            }
+            else
+            {
+                x = data[4];
+                y = data[5];
+                PlaySE(0x24);
```
Comment (jiangzhengwenjz) [inline src/field_tasks.c]: check include/constants/songs.h for corresponding macro to use

File: src/field_tasks.c [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+static void DummyPerStepCallback(u8 taskId)
+{
+}
+
+static void MarkIcefallCaveCoordVisited(s16 x, s16 y)
+{
+    u8 i = 0;
+    for (; i < NELEMS(sIcefallCaveIceTileCoords); ++i)
+    {
+        if (sIcefallCaveIceTileCoords[i][0] + 7 == x && sIcefallCaveIceTileCoords[i][1] + 7 == y)
```
Comment (mrgriffin) [inline src/field_tasks.c]: Is there a constant for the `+ 7` here and in `Special_SetIcefallCaveCrackedIceMetatiles`?

File: src/field_tasks.c [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+                data[6]--;
+            }
+            else
+            {
+                x = data[4];
+                y = data[5];
+                PlaySE(SE_RU_GASYAN);
+                MapGridSetMetatileIdAt(x, y, 0x35b);
+                CurrentMapDrawMetatileAt(x, y);
+                VarSet(0x4001, 1);
```
Comment (jiangzhengwenjz) [inline src/field_tasks.c]: sorry I missed this one. 
use constant defined in include/constants/vars.h

File: src/field_tasks.c (hunk 6 above)
Comment (PikalaxALT) [reply src/field_tasks.c]: Not that I'm aware of.

**Pull request merged**